File: bio.cpp


#include <iostream>
#include <math.h>
#include <stdint.h>
#include <string>
using namespace std;

typedef struct Float3
{
    float x;
    float y;
    float z;
} float3;

float my_nextafterf(float a, int dir);
float onemove_in_cube(float3 *p0, float3 *v, float *htime, int *id);
void printFloat(float3 *pos, string name);
void readArr(float (*x)[208][256][225]);
int main()
{
    const uint8_t Nx = 208;                                                             //Number of voxels in x direction of target
    const uint16_t Ny = 256;                                                            //Number of voxels in y direction of target
    const uint8_t Nz = 225;                                                             //Number of voxels in z direction of target
    const uint8_t Mx = 200;                                                             //Number of pixels in x direction of detector
    const uint8_t My = 200;                                                             //Number of pixels in y direction of detector
    const uint8_t D = 2;                                                                //Size in mm of each pixel side
    const uint8_t h = 50;                                                               //height above detector of target
    const uint16_t H = h + Nz + 200;                                                    //height to reach detector
    const float3 orginOffset = {(-Mx * D) / 2 + (Nx / 2), (-My * D) / 2 + (Ny / 2), 0}; //offset for orgin of detector
    const float3 ep = {Nx / 2, Ny / 2, H};                                              //location of detector
    const float muBone = 0.573;                                                         //mu for bone in scan
    const float muFat = 0.193;                                                          //my for fat in scan
    float mu[Nx][Ny][Nz];                                                               //Mu array of voxels
    float detector[Mx][My];                                                             //Array of detector pixels
    float3 pos;                                                                         //position vector
    float3 dir;                                                                         //direction vector
    uint8_t i, j;                                                                       //loop counters for unrolled loop
    float L;                                                                            //energy count
    readArr(mu); 
    //cout<<mu[0][0][0];
    for (int ii = 0; ii < Mx * My; ii = ii + 1)
    {
        i = ii / Mx;
        j = ii - i * Mx;

        pos = {orginOffset.x + D * i, orginOffset.y + D * j, 0};
        dir = {ep.x - pos.x, ep.y - pos.y, ep.z - pos.z};
        L = sqrt(pow(dir.x, 2) + pow(dir.y, 2) + pow(dir.z, 2));
        dir.x = dir.x / L;
        dir.y = dir.y / L;
        dir.z = dir.z / L;
        L = 1;

        /*while (pos.z < h + Nz)
        {
            //onemove_in_cube(float3 * p0, float3 * v, float *htime, int *id)
            if (pos.x >= 0 && pos.x < Nx && pos.y >= 0 && pos.y < Ny && pos.z >= h && pos.z < h + Nz)
            {
                cout << "In the cube";
                L = L * exp(-1);
            }
        }
        detector[i][j] = L;
        */
    }
}

void readArr(float (*x)[256][225]){
    cout<<"Prefault I"<<endl;
    for(int i=0;i<208;i=i+1){
    for(int j=0;j<256;j=j+1){
	for(int z=0;z<225;z=z+1){
	x[i][j][z]=0;
	}
	}
	}
    //cout<<"x(0,0,0) "<<x[0][0][0]<<endl;
}

void printFloat(float3 *pos, string name)
{
    cout << name << ": " << pos->x << " " << pos->y << " " << pos->z << endl;
}

/**
Move photon to the first intersecting wall inside a 1x1x1mm cube
Input:
 p0: pointer to a float3 struct, {x,y,z}current x/y/z position of the ray
 v: pointer to a float3 struct, {vx,vy,vz}, direction unitary vector
Output:
 return value: the shortest distance to intersect with a voxel wall, in mm
 htime: pointer to a float[3] array, store the intersection position (x/y/z) of the first wall
 id: pointer to an integer, a flag: id=0: intersect with x-plane; id=1: y-plane; id=3: z-plane
*/
float onemove_in_cube(float3 *p0, float3 *v, float *htime, int *id)
{
    float dist, xi[3];
    //time-of-fly to hit the wall in each direction
    htime[0] = fabs((floor(p0->x) + (v->x > 0.f) - p0->x) / v->x);
    htime[1] = fabs((floor(p0->y) + (v->y > 0.f) - p0->y) / v->y);
    htime[2] = fabs((floor(p0->z) + (v->z > 0.f) - p0->z) / v->z);
    //get the direction with the smallest time-of-fly
    dist = fmin(fmin(htime[0], htime[1]), htime[2]);
    (*id) = (dist == htime[0] ? 0 : (dist == htime[1] ? 1 : 2));
    //p0 is inside, p is outside, move to the 1st intersection pt, now in the air side, to be corrected in the else block
    htime[0] = p0->x + dist * v->x;
    htime[1] = p0->y + dist * v->y;
    htime[2] = p0->z + dist * v->z;
    xi[0] = my_nextafterf((int)(htime[0] + 0.5f), (v->x > 0.f) - (v->x < 0.f)); // if using matlab, xi=round(htime)+eps(single(1.0))*(v>0 – v<0);
    xi[1] = my_nextafterf((int)(htime[1] + 0.5f), (v->y > 0.f) - (v->y < 0.f));
    xi[2] = my_nextafterf((int)(htime[2] + 0.5f), (v->z > 0.f) - (v->z < 0.f));
    if (*id == 0)
        htime[0] = xi[0];
    if (*id == 1)
        htime[1] = xi[1];
    if (*id == 2)
        htime[2] = xi[2];
    return dist;
}
float my_nextafterf(float a, int dir)
{
    union {
        float f;
        unsigned int i;
    } num;

    num.f = a + 1000.f;
    num.i += dir ^ (num.i & 0x80000000U);
    return num.f - 1000.f;
}

File: bio.py
'''
Code developed by Anthony Bisulco
Feburary 26th, 2018
This is a script to run an imaging scenario on a given imaging volume
with an aribitrary sized detector. Currently, the file that loads is 
a brain scan which is imaged. This code uses Numba accleration in order 
to obtain resonable imaging times. Additionally, a verifcation is placed
to make sure this code gives them same results as a developed matlab code. 
'''
import warnings
with warnings.catch_warnings():
    warnings.filterwarnings("ignore",category=FutureWarning)
    import h5py
import numpy as np
import math
import numba
try:
    profile  # throws an exception when profile isn't defined
except NameError:
    profile = lambda x: x

@numba.jit(nopython=True, nogil=True, cache=True)
def onemove_in_cube_true_numba(p0,v):   
    '''   
    This is a function that moves from a given position p0 in direction v to another cube in a 1x1x1mm setup
    Args:
        p0: np.array 1x3 start position (X,Y,Z)
        v: np.array 1x3 noramlized(1) direction vector (X,Y,Z)

    Returns:
        htime: np.array 1x3 next cube position (X,Y,Z)
        dist: uint distance to the next cube position

    '''
    htime=np.abs((np.floor(p0)-p0+(v>0))/v) #find distance vector to new position
    minLoc=np.argmin(htime) #find min distance location in htime
    dist=htime[minLoc] #find min distance
    htime=p0+dist*v #calculate new position estimate htime
    htime[minLoc]=round(htime[minLoc])+np.spacing(abs(htime[minLoc]))*np.sign(v[minLoc]) #Need this for rounding to next position
    return htime,dist

@numba.jit(nopython=True, nogil=True, cache=True, parallel=True)
def main_loop(Nx,Ny,Nz,Mx,My,D,h,orginOffset,ep,mu, detector):
    '''
    Ray tracing from end point to all pixels, calculates energy at every pixels
    Args:
        Nx: uint imaging volume length in x direction
        Ny: uint imaging volume length in y direction
        Nz: uint imaging volume length in z direction
        Mx: uint number of pixels in x direction
        My: uint number of pixels in y direction
        D: uint pixel length
        h: uint distance from detector to bottom of imaging volume
        orginOffset: np.array 1x2 offset origin for detector position start (X,Y)
        ep: np.array 1x3 location of the x-ray source (X,Y,Z)
        mu: np.array 1x3 noramlized linear attenuation coeffcient matrix (Nx,Ny,Nz)
    Returns:
        detector: np.array 1x3 next cube position (X,Y,Z)
    '''
    
    for z in numba.prange(Mx*My): #loop for all pixels
        j=z%Mx #y direction pixel
        i=int(z/Mx) #x direction pixel
        pos=np.array([orginOffset[0]+i*D,orginOffset[1]+D*j, 0],dtype=np.float32) #pixel location
        dir=((ep-pos)/np.linalg.norm(ep-pos)).astype(np.float32) #noramlized direction vector to source
        dir[dir==0]=1e-16 #need this for floating point division errors in Numba
        L=0 # initial energy
        h_z=h+Nz
        while pos[2]< h_z: #loop until the end of imaging volume
            pos,dist=onemove_in_cube_true_numba(pos,dir) #move to next cube
            if 0 <= pos[0] < Nx and 0<=pos[1]<Ny  and h<=pos[2] < h_z: #if in imaging volume
                L+=mu[math.floor(pos[0]),math.floor(pos[1]),math.floor(pos[2]-h)]*dist #calculate energy using mu
        detector[i][j] = L; # detector pixel locaiton equals lasting energy
    return detector
 

@profile
def main():
    f = h5py.File('headct.h5', 'r') #HDF5 file containing Headct array of linear attenuation coeffcients(Nx,Ny,Nz)
    headct=np.array(f.get('ct'))
    headct=np.transpose(headct) #linear attenuation coeffceient matrix
    det=f.get('det')
    det=np.transpose(det) #Result of running this code on Matlab for later comparsion
    Nx = np.size(headct,0) #Imaging x dimension length in mm
    Ny = np.size(headct,1) #Imaging y dimension length in mm
    Nz = np.size(headct,2) #Imaging z dimension length in mm
    Mx = 200 #Number of pixels in x direction
    My = 200 #Number of pixels in y direction
    D = 2 #Size of each pixel in mm
    h = 50 #distance(Z) bettween bottom of imaging volume and detector
    H = h + Nz + 200 #distance(Z) bettween detector and x-ray source
    muBone = 0.573 #linear attenuation coeffcient bone cm^-1
    muFat = 0.193 #linear attenuation coeffcient fat cm^-1
    orginOffset = np.array(
            [(-Mx * D) / 2 + (Nx / 2), (-My * D) / 2 + (Ny / 2), 0], dtype=np.float32) #offset from origin to detector start (X,Y,Z)
    ep = np.array([Nx / 2, Ny / 2, H], dtype = np.float32) #location of x-ray soruce
    orginOffset = np.array([(-Mx * D) / 2 + (Nx / 2), (-My * D) / 2 +(Ny / 2), 0],dtype=np.float32) #offset from origin to detector start
    mu=np.zeros((Nx,Ny,Nz),dtype=np.float32) #(Nx,Ny,Nz) linear attenuation coeffcient matrix 
    mu[np.nonzero(headct>0)]=((headct[np.nonzero(headct>0)]-0.3)/(0.7))*(muBone-muFat)+muFat #Normilization of givens mus of linear attenuation matrix
    detector=np.zeros((Mx,My),dtype=np.float32) #detector Mx x pixels and My y pixels
    detA=main_loop(Nx,Ny,Nz,Mx,My,D,h,orginOffset,ep,mu,detector)
    detector=np.exp(detA*-10,dtype=np.float64)
    print(np.count_nonzero(np.isclose(det,detector,rtol=0.01))>39900)
    np.save('Detector.npy',detector)

if __name__ == '__main__':
    main()
